### PR TITLE
:sparkles: pkg/crd: support validating internal list items on list types

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
For #342

### CODE REVIEW

**Comment by AlexanderYastrebov** [comment]
Maybe zz_generated.markerhelp.go files should be updated first in a separate PR to reduce the diff.

File: pkg/crd/markers/validation.go [hunk 1]
```diff
@@ -53,6 +53,10 @@ var ValidationMarkers = mustMakeAllWithPrefix("kubebuilder:validation", markers.
 	MinLength(0),
 	Pattern(""),
 
+	ItemMaxLength(0),
+	ItemMinLength(0),
+	ItemPattern(""),
```
Comment (JoelSpeed) [inline pkg/crd/markers/validation.go:58]: What about XValidation or Enum? And how do we want to handle slice of int for example, min and max values there would also be required for a complete solution.

Wondering if we need to support most of the current validations that can apply to a built-in as items versions as well

Comment (AlexanderYastrebov) [reply pkg/crd/markers/validation.go:58]: Yes, for a complete solution most of all scalar markers should have item variant.
If proposed approach is fine in general I can update the PR accordingly.

File: pkg/crd/testdata/cronjob_types.go [hunk 2]
```diff
@@ -360,6 +378,12 @@ type LongerString string
 // TotallyABool is a bool that serializes as a string.
 type TotallyABool bool
 
+// This tests string slice item validation.
+// +kubebuilder:validation:ItemMinLength=1
+// +kubebuilder:validation:ItemMaxLength=255
+// +kubebuilder:validation:ItemPattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
```
Comment (alvaroaleman) [inline pkg/crd/testdata/cronjob_types.go:384]: Would it be possible to do something like this:
```suggestion
// +kubebuilder:validation:items:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
```

and then internally  re-use the existing validations rather than duplicating all of them?

Comment (AlexanderYastrebov) [reply pkg/crd/testdata/cronjob_types.go:384]: Ok, I've created #899

**Comment by AlexanderYastrebov** [comment]
Closing in favour of #899

File: pkg/crd/testdata/cronjob_types.go (hunk 2 above)
Comment (alvaroaleman) [reply pkg/crd/testdata/cronjob_types.go:384]: @AlexanderYastrebov can you please update your existing PR rather than creating a new one for every change? Otherwise the whole discussion is spread up among a ton of PR and hard to follow-up on later on.

Comment (AlexanderYastrebov) [reply pkg/crd/testdata/cronjob_types.go:384]: I closed this and previous as they use different approaches.

Comment (alvaroaleman) [reply pkg/crd/testdata/cronjob_types.go:384]: @AlexanderYastrebov that doesn't matter, please update the existing PR so there is one place that has the history of the discussion

Comment (AlexanderYastrebov) [reply pkg/crd/testdata/cronjob_types.go:384]: I can not reopen this closed PR fro some reason, I'll take a note for the next time I contribute.

Comment (AlexanderYastrebov) [reply pkg/crd/testdata/cronjob_types.go:384]: Ok, I think I managed to reopen it.

File: pkg/crd/markers/validation.go (hunk 1 above)
Comment (AlexanderYastrebov) [reply pkg/crd/markers/validation.go]: I've updated the PR to support all types of validation markers.

File: pkg/crd/testdata/cronjob_types.go [hunk 3]
```diff
@@ -252,6 +252,24 @@ type CronJobSpec struct {
...
+
+	// This tests string alias slice item validation.
+	LongerStringArray []LongerString `json:"longerStringArray,omitempty"`
+
+	// This tests that a slice of IntOrString can also have a pattern attached to it.
+	// This can be useful if you want to limit the string to a perecentage or integer.
+	// The XIntOrString marker is a requirement for having a pattern on this type.
+	// +kubebuilder:validation:items:XIntOrString
+	// +kubebuilder:validation:items:Pattern="^((100|[0-9]{1,2})%|[0-9]+)$"
+	IntOrStringArrayWithAPattern []*intstr.IntOrString `json:"intOrStringArrayWithAPattern,omitempty"`
```
Comment (alvaroaleman) [inline pkg/crd/testdata/cronjob_types.go:272]: Could you also add some testcases for validations on slices that validate the slice as a whole instead of `items`, please? I will try to review this over the weekend

Comment (AlexanderYastrebov) [reply pkg/crd/testdata/cronjob_types.go:280]: Added `kubebuilder:validation:MinItems` to `Hosts` to test the mix and also added `StringPair` test.

**Comment by JoelSpeed** [comment]
Seems a lot cleaner than the original approach, and much less ambiguous as well, thanks for taking on board the UX feedback

**Review by alvaroaleman — APPROVED** [review]
looks good me, thanks for making this happen!
/hold
So @vincepri and @sbueringer get a chance to have a look

**Comment by sbueringer** [comment]
Thank you very much!

/lgtm

**Comment by AlexanderYastrebov** [comment]
@alvaroaleman @sbueringer Thank you 🙏
